_posts/2021-12-09-kafka-segment-retention.md

```diff
@@ -0,0 +1,259 @@
+The first condition that is met will trigger the rolling of a new segment.
+It is also worth noticing that, as a not so usual use case, the producer timestamps in the records could be not ordered, so not having the older record as the first one, due to retries or the specific business logic of the producer.
+
+Another useful parameter is `log.roll.jitter.ms` which sets a maximum jitter when it's time to roll a segment,. Jitters are time intervals to avoid many segments being rolled at same time causing high contention on the disk.
+A jitter is randomly generated for each segment.
+
+The above conditions, by size or time, are the well known ones but not everyone knows that there is a third condition.
+
+A segment is rolled even when the corresponding index (or timeindex) is full.
+The index and timeindex share the same maximum size which is defined by the `log.index.size.max.bytes` configuration parameter and it is 10 MB by default.
+Let's consider the default log segment maximum size, which is 1 GiB.
+We know that because `log.index.interval.bytes` is 4096 bytes by default, an entry is added in the index every 4096 bytes of records.
+It means that for a 1 GiB segment size, 1 GiB / 4096 bytes = 262144 entries are added to the index. This equals 2 MB of index (262144 * 8 bytes).
+The default index size of 10 MB is enough to handle a segment size of 5 GiB.
+This means that reducing the index size or increasing the segment size will mean a new segment is rolled when the index is full and not when the requested segment size is reached.
+So, setting a segment size larger than 5 GiB without increasing the index size is useless, because the broker will roll a new segment when the index is full.
+
+By increasing the segment size over 5 GiB, you would also need to increase the index file size as well. Likewise, if you decide to reduce the index file size, you should probably decrease the segment size as well.
+
+The timeindex might need attention as well. Because each timeindex entry is 1.5x bigger than an entry in the index (12 bytes versus 8 bytes), it can fill up earlier and cause a new segment to be rolled.
+
+To show how the index and timeindex files size have an impact on rolling a new log segment, let's consider a cluster with `log.index.interval.bytes=150` and `log.index.size.max.bytes=300` still using the Strimzi Canary to produce and consume records with the usual configuration `retention.ms=600000;segment.bytes=16384`.
+Because the Strimzi Canary records are close to 150 bytes in size, we would expect the index file to be filled with one entry every two records.
+With the maximum size set at 300 bytes, we would have around 300 / 8 = 37 entries in the index file and around 300 / 12 = 25 entries in the timeindex file.
```

Comment:
These last 2 paragraphs are not easy to follow. 
What about something like this:
- say the timeindex is the index that can cause roll because it uses a longer keys than other index as described above
- compute how many entries it can contain with the size set
- extrapolate how many records this maps to
- look at example to confirm roll happened before segment was full due to index filling up first

Reply:
> say the timeindex is the index that can cause roll because it uses a longer keys than other index as described above

I think I addressed it on line 178 with "Because each timeindex entry is 1.5x bigger than an entry in the index (12 bytes versus 8 bytes), it can fill up earlier and cause a new segment to be rolled."

> extrapolate how many records this maps to

Not sure what you mean there is no references to the number of records. I am not mentioning them at all, is that important? I said something like "index file to be filled with one entry every two records." ... maybe this information is not useful at all in this context. We don't want to explain in terms of records but just highlighting that the log segment was rolled before 16k because of the index/timeindex full.

Reply:
No worries, that was just a suggestion.